### ISSUE TITLE
SGE Java DRMAA not working

### ISSUE THREAD

**Issue by tkainrad** [aws/aws-parallelcluster#190]
Using CfnCluster 1.3.1, the DRMAA Java Bindings for SGE do not work.
I have tested with multiple OS types (alinux, centos, ubuntu). 
I am used to use the Java DRMAA Bindings with a local Cluster running SGE 8.1.8 (CfnCluster uses 8.1.9). 
There it is enough to pass the option
`-Djava.library.path=/opt/sge/lib/lx-amd64/`
to the command line.

However, with CfnCluster I always get
`nested exception is java.lang.UnsatisfiedLinkError: com.sun.grid.drmaa.SessionImpl.nativeInit`

Apparently drmaa.jar in /opt/sge/lib/ is missing, which may be part of the problem. However, just copying the 8.1.9 drmaa.jar from https://arc.liv.ac.uk/trac/SGE does not solve the issue.

**Comment by KainiT** [comment]
I have now identified the problem. SGE is explicitly compiled without java support:

See line 38 in file [https://github.com/awslabs/cfncluster-cookbook/blob/master/recipes/sge_install.rb](url)

Why is this? Would it be possible to change this or provide configuration options to prevent it?

In case anyone has the same problem, I managed to get this working using the following workaround:
Compile Son of Grid Engine on the same architecture and OS as the CfnCluster nodes are using. In my case this was Ubuntu 16.04 amd64. Use the resulting /opt/sge/lib folder to replace the one that is present on the Cloud Cluster. In the case of Ubuntu 16.04 this lead to the native library (libdrmaa.so) missing libjemalloc. This can then be fixed by simply installing it:
`sudo apt-get install libjemalloc-dev`

The above workaround is not very clean to script via post_install, also it does not work the same on all OS types, making it necessary to use different EBS snapshots upon cluster creation. Therefore, I would really appreciate the possibility to have SGE compile with Java support in the first place.

**Comment by bwbarrett** [comment]
You've properly diagnosed the problem; we don't compile Java support into SGE.  Adding such support is not currently in our roadmap, but we'll keep this issue open to track the request.  Do you have a use case that could help us understand the need?

**Comment by KainiT** [comment]
I suppose all DRMAA Use Cases apply to this. 
I have a very classical one: I develop a Java server application that lets client programs submit jobs to (for now) SGE clusters using web services. I am intrigued by CfnCluster and very interested in supporting it as well as possible. The alternative to DRMAA is of course having the server-side application execute qsub, qstat,... commands manually which requires lots of error-prone output parsing. I will likely have to do it anyways at some point in order to support other grid engines besides SGE. However, from a more general point of view, I think CfnCluster is even more well suited for this kind of remote execution servers than on-site clusters. Supporting technologies that enable and/or simplify these scenarios  seems worthwile to me.

Keep up the good work, I and many others I talked to see grid computing with job scheduling as a perfect cloud computing use case.

**Comment by bwbarrett** [comment]
Thanks for the extra information.  Not sure when we'll be able to prioritize the work, but it's on our request list.

**Comment by CharlesEGrant** [comment]
I'm another user that's being blocked by this limitation. We run a [public web site](http://meme-suite.org) for bioinformatics analysis. Currently we're using [StarCluster](http://star.mit.edu/cluster/), but that has not been under active development for some time. We'd love to switch to AWS ParallelCluster, but our site is designed around Tomcat and [Opal](https://sourceforge.net/projects/opaltoolkit/], but the latter submits jobs to the SGE queue using the Java DRMAA library. We'd be very grateful if support for the Java DRMAA could be added to the SGE build.

**Comment by tkainrad** [comment]
I would also still appreciate this enhancement.

However, I have found a workaround that I use in-production and it has been working for us without problems for several months now. 

Basically, I did the following:
1.  Compile SGE locally with Java support
2. store the resulting `/opt/sge/lib/` folder in the EBS snaphots used with AWS Parallelcluster
3. on cluster startup (`post-install` script), replace the default `/opt/sge/lib/` folder with the one saved in the AWS snapshot
4. Depending on your operating system, you might need to install some additional libraries, with Ubuntu, we need to do `sudo apt-get install libjemalloc-dev -y`

Let me know if you have further questions on this.

**Comment by nstornetta** [comment]
Because we have announced that we will be deprecating support for SGE in the near-future (see: https://github.com/aws/aws-parallelcluster/wiki/Deprecation-of-SGE-and-Torque-in-ParallelCluster), we will not be performing additional enhancements specific to SGE.

I am going to close this issue. If you would like to request a similar enhancement for one of our other supported schedulers (Slurm or AWS Batch), please feel free to create a new issue.
